lucene/core/src/test/org/apache/lucene/search/TestConstantScoreScorer.java

```diff
@@ -0,0 +1,104 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.search;
+
+import java.io.IOException;
+
+import org.apache.lucene.document.Document;
+import org.apache.lucene.document.Field;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.RandomIndexWriter;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.util.LuceneTestCase;
+
+import static org.apache.lucene.search.BooleanClause.Occur;
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.is;
+
+public class TestConstantScoreScorer extends LuceneTestCase {
+  private static final String FIELD = "f";
+  private static final String[] VALUES = new String[]{"foo", "bar", "foo bar", "azerty"};
+
+  public void testBasics() throws Exception {
+    try (final TestConstantScoreScorerIndex index = new TestConstantScoreScorerIndex()) {
+      BooleanQuery query = new BooleanQuery.Builder()
+          .add(new ConstantScoreQuery(new TermQuery(new Term(FIELD, "foo"))), Occur.MUST)
+          .build();
+
+      final ConstantScoreScorer scorer = index.constantScoreScorer(query, 1f);
+      assertThat(scorer.iterator().nextDoc(), equalTo(0));
+      assertThat(scorer.score(), is(equalTo(1f)));
+      assertThat(scorer.iterator().nextDoc(), equalTo(2));
+      assertThat(scorer.score(), is(equalTo(1f)));
+      assertThat(scorer.iterator().nextDoc(), equalTo(NO_MORE_DOCS));
+    }
+  }
+
+  public void testWithMinCompetitiveScoreSet() throws Exception {
+    try (final TestConstantScoreScorerIndex index = new TestConstantScoreScorerIndex()) {
+      BooleanQuery query = new BooleanQuery.Builder()
+          .add(new ConstantScoreQuery(new TermQuery(new Term(FIELD, "foo"))), Occur.MUST)
+          .build();
+
+      final ConstantScoreScorer scorer = index.constantScoreScorer(query, 1f);
+      assertThat(scorer.iterator().nextDoc(), equalTo(0));
+      assertThat(scorer.score(), is(equalTo(1f)));
+      scorer.setMinCompetitiveScore(2f);
+      assertThat(scorer.iterator().nextDoc(), equalTo(NO_MORE_DOCS));
+    }
+  }
+
+  static class TestConstantScoreScorerIndex implements AutoCloseable {
+
+    private final Directory directory;
+    private final RandomIndexWriter writer;
+    private final IndexReader reader;
+
+    TestConstantScoreScorerIndex() throws IOException {
+      directory = newDirectory();
+
+      writer = new RandomIndexWriter(random(), directory);
```

Comment:
you need to use a merger that retains doc id order. Something like:
```
RandomIndexWriter w = new RandomIndexWriter(random(), dir, newIndexWriterConfig().setMergePolicy(
        // retain doc id order
        newLogMergePolicy(random().nextBoolean())));
```

Reply:
Done in 0ad2f781790f35c6b7f16531db182b743d13a7c0, thank you.